### ISSUE TITLE
Revert removal of pixel snapping from flutter engine

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#111145]
See https://github.com/flutter/flutter/issues/110959, https://github.com/flutter/flutter/issues/110957, likely others.

The assumption of the raster cache is that we can replace any layer or display list with a rasterized version of itself. At very low DPR, this is only true if that layer or display list is aligned to physical pixels. Even if we ensure that the texture of the layer or display list is rasterized pixel aligned, when we go to the draw that entry we may find the offset is not physical pixel aligned. Thus we can either snap it to a pixel, leaving it in a different location than the un-rasterized version, or we can draw it in the correct location with obvious aliasing issues.

Unfortunately, that leaves us in a bit of a bind. It means we can't get rid of pixel snapping as long as we have the raster cache. Furthermore, since we didn't notice this issue until the stable release, and there have been many refactors on top of the raster cache since then, it means that a revert is not very straight forward.

Seemingly the only idea long term solution is to revert back to pixel snapping any layers that may be raster cached. The short term change is more uncertain. I can begin working on a patch against the release branch that reverts the removal of pixel snapping. Separately I can also work on re-landing pixel snapping on ToT.

@CaseyHillers do you have a stable release engine branch I can use to investigate the feasibility of a targeted revert?

FYI @zanderso

**Comment by zanderso** [comment]
> we can draw it in the correct location with obvious aliasing issues.

Could you describe what these issues are?

**Comment by jonahwilliams** [comment]
Watch as the tooltip fades in. This is the logic for "draw it in the same place it would be drawn". Notice how the text looks aliased. Once the transition is complete, the raster cache entry is dropped and we draw it normally. Skia must internally have better anti-aliasing logic for fonts if it knows it is being drawn at a fractional pixel vs if its is drawn pixel aligned into a texture which is then composited into an app.

https://user-images.githubusercontent.com/8975114/188994983-b25c4fe9-69e8-47ca-81d7-50a2b0bfbac6.mp4

For high dpr screens, we can get away with just snapping the pixel. For low dpr screens it leads to noticeable movement (which is the whole issue that led me down the path)

**Comment by CaseyHillers** [comment]
Current stable is in https://github.com/flutter/engine/tree/flutter-3.2-candidate.5

For future reference, you can go to https://github.com/flutter/flutter/blob/stable/bin/internal/release-candidate-branch.version to find the current candidate branch used by stable.

The existing stable branch is two months old, and IIRC this was very difficult to get rolled into Google as well.

**Comment by zanderso** [comment]
Wondering if @flar might have any thoughts. I understand @jonahwilliams has spent a fair amount of time thinking about this, but at least on master, we are not under time pressure to come up with a quick solution.

**Comment by jonahwilliams** [comment]
I'm not aware of this change having any impact in google3. @CaseyHillers what made it difficult to land?

**Comment by jonahwilliams** [comment]
Okay, I've concluded my investigation of the release branch. Because the removal of pixel snapping occurred before a raster cache refactor and a display_list layer refactor, there is no clean revert without reverting quite a long chain of other PRs. Furthermore, it isn't sufficient to simply draw the cache entries at pixel snapped positions, since this will conflict with the reported positions for layer diffing.

So for stable channel, the potential options are reverting 10+ commits in order to keep things mostly inline, or to create a new PR against stable that re-introduces pixel snapping to the raster cached layers.

**Comment by CaseyHillers** [comment]
@jonahwilliams I followed up. I misremembered this with https://github.com/flutter/flutter/commit/fba9214

\cc @itsjustkevin

**Comment by jonahwilliams** [comment]
Ahh yeah, I promise we don't do that again. And with this revert, we will still _not_ have pixel snapping in the tester, which means the google3 goldens will remain somewhat stabilized

**Comment by itsjustkevin** [comment]
@jonahwilliams my question here is, "is the juice worth the squeeze?"

With the branch being two months old and the roll state not in an ideal place, can we consider doing this work at another time in the near future?

A chain of 10+ reverts seems inherently risky.

**Comment by jonahwilliams** [comment]
Yes, the juice is worth the squeeze.

https://github.com/flutter/engine/pull/35981 adds back the pixel snapping, and is not as disurptive as I was worried it to be, though a straight revert is infeasible.

**Comment by knopp** [comment]
@jonahwilliams, depending on how close the snapping behavior is to before removal this one might need to be reverted too. https://github.com/flutter/engine/pull/35601 :-/
